File: dot_prod/dot_prod_omp_simple.c
/* Daniel R. Reynolds
   SMU Mathematics
   Math 6370
   14 January 2013 */


/* Inclusions */
#include <stdlib.h>
#include <stdio.h>
#include "get_time.h"



/* Example routine to compute the dot-product of two vectors. */
int main(int argc, char* argv[]) {

  /* declarations */
  int i, n;
  double *a, *b, sum, alloctime, inittime, runtime;
  double stime, ftime;

  /* ensure that an argument was passed in */
  if (argc < 2) {
    printf("Error: function requires one argument (vector length)\n");
    return 1;
  }

  /* set n as the input argument, and ensure it's positive */
  n = atoi(argv[1]);
  if (n < 1) {
    printf("Error: vector length %i must be greater than 0\n", n);
    return 1;
  }

  /* allocate the vectors */
  stime = get_time();
  a = malloc(n*sizeof(double));
  b = malloc(n*sizeof(double));
  ftime = get_time();
  alloctime = ftime - stime;

  /* initialize the vector values */
  stime = get_time();
  for (i=0; i<n; i++)    a[i] = (0.001 * (i + 1.0)) / n;
  for (i=0; i<n; i++)    b[i] = (0.001 * (n - i - 1.0)) / n;
  ftime = get_time();
  inittime = ftime - stime;

  /* compute dot-product */
  stime = get_time();
  sum = 0.0;
  #pragma omp parallel for reduction(+:sum)
  for (i=0; i<n; i++)  sum += a[i]*b[i];
  ftime = get_time();
  runtime = ftime - stime;

  /* output computed value and runtime */
  printf(" vector length = %i\n",n);
#ifdef _OPENMP
  printf("   num threads = %i\n", omp_get_num_threads());
#endif
  printf("   dot-product = %.16e\n",sum);
  printf("    alloc time = %.2e\n",alloctime);
  printf("     init time = %.2e\n",inittime);
  printf("      run time = %.2e\n",runtime);

  /* free vectors */
  free(a);
  free(b);

  return 0;
} /* end main */

File: dot_prod/dot_prod_omp_simple.f90
! -*- Mode: Fortran90; -*-
!-----------------------------------------------------------------
! Daniel R. Reynolds
! SMU, Mathematics
! Math 6370
! 13 January 2013
!=================================================================


program DotProd_OpenMP
  !-----------------------------------------------------------------
  ! Description: 
  !    Computes the dot-product of two vectors.
  !-----------------------------------------------------------------
  !======= Inclusions ===========

  !======= Declarations =========
  implicit none

  integer :: i, n
  double precision, allocatable :: a(:), b(:)
  double precision :: sum, alloctime, inittime, runtime
  double precision :: stime, ftime
  double precision, external :: omp_get_wtime
  
  !======= Internals ============
  
  ! get the vector length from the user
  print *, 'input vector length n (must be >0):'
  read(*,*) n

  ! ensure that n is positive
  if (n < 1) then
     print *, 'Error: vector length', n, ' must be greater than 0'
     stop
  endif

  ! alloctate the vectors
  stime = omp_get_wtime()
  allocate(a(n),b(n))
  ftime = omp_get_wtime()
  alloctime = ftime-stime

  ! initialize the vectors
  stime = omp_get_wtime()
  do i=1,n
     a(i) = 1.d-3*i/n
  enddo
  do i=1,n
     b(i) = 1.d-3*(n-i)/n
  enddo
  ftime = omp_get_wtime()
  inittime = ftime-stime

  ! perform parallel dot-product
  stime = omp_get_wtime()
  sum = 0.d0
  !$omp parallel do reduction(+:sum)
  do i=1,n,1
     sum = sum + a(i)*b(i)
  enddo
  !$omp end parallel do
  ftime = omp_get_wtime()
  runtime = ftime-stime

  ! output computed value and runtime
  print *, ' vector length =',n
  print *, '   dot-product =',sum
  print *, '    alloc time =',alloctime
  print *, '     init time =',inittime
  print *, '      run time =',runtime

  ! free vectors
  deallocate(a,b)

end program DotProd_OpenMP
!=================================================================
